### PR TITLE
[NativeAOT] Making thread local storage completely managed.

### PR DESCRIPTION
**Pull request by VSadov**
Basically implementing `m_pThreadLocalModuleStatics` as a managed `object[][]` array

- no need to involve unmanaged allocator and GC handles, thus access to threadstatics has one less indirection
- managing the threadlocal storage happens entirely in managed code
- destruction of the entire threadstatics structure is a noop

The threadstatics root is reported together with other thread-owned roots like locals. 
Once the owning thread's lifecycle gets to the stage when it stops participating in root reporting, `m_pThreadLocalModuleStatics` can be reclaimed.

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/thread.h [hunk 1]
```diff
@@ -88,8 +88,7 @@ struct ThreadBuffer
     uintptr_t               m_uHijackedReturnValueFlags;            
     PTR_ExInfo              m_pExInfoStackHead;
     Object*                 m_threadAbortException;                 // ThreadAbortException instance -set only during thread abort
-    PTR_PTR_VOID            m_pThreadLocalModuleStatics;
-    uint32_t                m_numThreadLocalModuleStatics;
+    Object*                 m_pThreadLocalModuleStatics;
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.h:91]: Do we need to report this to the GC now that it is a raw `Object*` ?

File: src/coreclr/nativeaot/Runtime/gcrhscan.cpp [hunk 2]
```diff
@@ -53,6 +54,9 @@ void GCToEEInterface::GcScanRoots(EnumGcRefCallbackFunc * fn,  int condemned, in
         else
 #endif
         {
+            STRESS_LOG1(LF_GC | LF_GCROOTS, LL_INFO100, "{ Scanning Thread's %p thread statics root. \n", pThread);
+            GcEnumObject(pThread->GetThreadStaticStorage(), 0 /*flags*/, fn, sc);
```
Comment (VSadov) [inline src/coreclr/nativeaot/Runtime/gcrhscan.cpp:58]: Here we report the location of threadstatic root to the GC - to keep the array alive and the pointer updated as long as the owner thread is alive. 
Just another root reported by the thread.

File: src/coreclr/nativeaot/Runtime/thread.h (hunk 1 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.h:91]: Yes it is reported as one of the roots that belong to the thread.

**Comment by VSadov** [comment]
closed/reopened to unstuck the CLA check

**Comment by VSadov** [comment]
Assert in ilc  (I think unrelated to this change)

```
 ILC: D:\a\_work\1\s\src\coreclr\jit\lir.cpp:1691
  ILC: Assertion failed 'node->IsUnusedValue() && "found an unmarked unused value"' in 'System.Net.Sockets.Tests.ConnectEap+<>c__DisplayClass1_0+<<ConnectAsync_WithData_DataReceived>b__0>d:MoveNext():this' during 'Lowering nodeinfo' (IL size 473; hash 0x4698aa3f; FullOpts)
 
```

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by VSadov** [comment]
the ilc assert is https://github.com/dotnet/runtime/issues/81356

**Comment by VSadov** [comment]
Thanks!

**Comment by MichalStrehovsky** [comment]
> the ilc assert is https://github.com/dotnet/runtime/issues/81356

That issue is closed. Maybe same assert, but definitely a different root cause.

Could you create an issue and exclude the test from building like I did in #83167 and #83250 (PR that added exclusion when I first noticed it).

We should not be merging PRs that have an entire test leg on the floor because of a build failure.

**Comment by VSadov** [comment]
Right. It is a slightly different assert. It needs a recent build but happens without my change. 
(a testrun with the commit just prior to mine reproes the failure, but it might have been introduced earlier)

The failure is only in Chk builds, so who does not run "extra-platforms" will not see the failure.

**Comment by VSadov** [comment]
The assert looked very similar at first glance and I thought it was fixed since my run hit it. I did not notice that the issue is fairly old.

**Pull request merged**